### PR TITLE
[wip]Reworks/Readds Fusion

### PR DESCRIPTION
**Pull request by as334**
:cl: optional name here
add: Fusion has be reworked. It now requires carbon dioxide to catalyze the plasma in the reaction.
/:cl:

I wasn't really happen with how fusion worked before(especially that it was broken) so this is my attempted reworking of it. 
New fusion uses plasma and carbon dioxide, with fusion firing in pulses that use up all available carbon dioxide(requires refilling the reaction chamber somehow). This is theoretically merge-able, but needs testing and I want to add functionality and interactions with the other gases.

### CODE REVIEW

File: code/modules/atmospherics/gasmixtures/reactions.dm [hunk 1]
```diff
@@ -182,7 +182,7 @@
 
 	return cached_results[id] ? REACTING : NO_REACTION
 
-//fusion: a terrible idea that was fun but broken. Now reworked to be less broken and more interesting.
+//fusion: a terrible idea that was fun but broken. Now reworked to be less broken and more interesting. Again.
```
Comment (duncathan) [inline code/modules/atmospherics/gasmixtures/reactions.dm]: it keeps happening

File: code/modules/atmospherics/gasmixtures/reactions.dm [hunk 2]
```diff
@@ -193,44 +193,43 @@
...
-	cached_gases[/datum/gas/plasma][MOLES] = max(cached_gases[/datum/gas/plasma][MOLES] - plasma_fused,0)
-	cached_gases[/datum/gas/tritium][MOLES] = max(cached_gases[/datum/gas/tritium][MOLES] - tritium_catalyzed,0)
-	cached_gases[/datum/gas/oxygen][MOLES] += oxygen_added
-	cached_gases[/datum/gas/water_vapor][MOLES] += waste_added
-	cached_gases[/datum/gas/nitrous_oxide][MOLES] += waste_added
-	cached_gases[/datum/gas/nitryl][MOLES] += waste_added
-	cached_gases[/datum/gas/carbon_dioxide][MOLES] += waste_added
-	if (location)
-		radiation_pulse(location, reaction_energy/(PLASMA_BINDING_ENERGY*MAX_CATALYST_EFFICENCY))
+	var/mediation = (air.heat_capacity()-(cached_gases[/datum/gas/plasma][MOLES]*cached_gases[/datum/gas/plasma][GAS_META][META_GAS_SPECIFIC_HEAT]))/(air.total_moles()-cached_gases[/datum/gas/plasma][MOLES])
```
Comment (duncathan) [inline code/modules/atmospherics/gasmixtures/reactions.dm]: this is temperate in the gases other than plasma, correct? could you indicate that with a comment?

File: code/modules/atmospherics/gasmixtures/reactions.dm [hunk 3]
```diff
@@ -193,44 +193,43 @@
...
+	var/carbon_amount = 2*cached_gases[/datum/gas/carbon_dioxide][MOLES]
```
Comment (duncathan) [inline code/modules/atmospherics/gasmixtures/reactions.dm]: why 2

File: code/modules/atmospherics/gasmixtures/reactions.dm [hunk 4]
```diff
@@ -193,44 +193,43 @@
...
+	var/plasma_fused = 0
+	if (carbon_amount < 20) //Lack of catalyst, fusion reaction starts to break down.
+		plasma_fused = carbon_amount
+		reaction_energy += plasma_fused*PLASMA_BINDING_ENERGY
+		cached_gases[/datum/gas/carbon_dioxide][MOLES] -= carbon_amount/20
+		air.assert_gases(/datum/gas/bz,/datum/gas/nitrous_oxide)
+		cached_gases[/datum/gas/bz][MOLES] += cached_gases[/datum/gas/carbon_dioxide][MOLES]/20
+		cached_gases[/datum/gas/nitrous_oxide][MOLES] += cached_gases[/datum/gas/carbon_dioxide][MOLES]/20
```
Comment (duncathan) [inline code/modules/atmospherics/gasmixtures/reactions.dm]: after you remove the co2, update `carbon_amount` and use it instead of checking moles again

File: code/modules/atmospherics/gasmixtures/reactions.dm [hunk 5]
```diff
@@ -193,44 +193,43 @@
...
+		if (location)
+			radiation_pulse(location, reaction_energy/(PLASMA_BINDING_ENERGY))
+	else
+		plasma_fused = max((carbon_amount*5),MAX_CATALYST_EFFICENCY)*((cached_gases[/datum/gas/plasma][MOLES])/mediation)
```
Comment (duncathan) [inline code/modules/atmospherics/gasmixtures/reactions.dm]: mol*mol/K = mol?

File: code/modules/atmospherics/gasmixtures/reactions.dm [hunk 6]
```diff
@@ -193,44 +193,43 @@
...
+		reaction_energy += plasma_fused*PLASMA_BINDING_ENERGY
```
Comment (duncathan) [inline code/modules/atmospherics/gasmixtures/reactions.dm]: just to verify, is `PLASMA_BINDING_ENERGY` in J/mol?

File: code/modules/atmospherics/gasmixtures/reactions.dm [hunk 7]
```diff
@@ -193,44 +193,43 @@
...
+		air.assert_gases(/datum/gas/oxygen)
+		cached_gases[/datum/gas/plasma][MOLES] -= plasma_fused
+		cached_gases[/datum/gas/carbon_dioxide][MOLES] = 0
```
Comment (duncathan) [inline code/modules/atmospherics/gasmixtures/reactions.dm]: this will make fusion only tick once without extra co2 being added. is this intentional?

File: code/modules/atmospherics/gasmixtures/reactions.dm [hunk 8]
```diff
@@ -254,7 +253,8 @@
 	var/heat_efficency = temperature/(FIRE_MINIMUM_TEMPERATURE_TO_EXIST*100)
 	var/energy_used = heat_efficency*NITRYL_FORMATION_ENERGY
 	ASSERT_GAS(/datum/gas/nitryl,air)
-
+	if ((cached_gases[/datum/gas/oxygen][MOLES] - heat_efficency < 0 )|| (cached_gases[/datum/gas/nitrogen][MOLES] - heat_efficency < 0)) //Shouldn't produce gas from nothing.
```
Comment (duncathan) [inline code/modules/atmospherics/gasmixtures/reactions.dm]: shouldn't these be `<=` checks?

Comment (duncathan) [reply code/modules/atmospherics/gasmixtures/reactions.dm]: and similar checks elsewhere

File: code/modules/atmospherics/gasmixtures/reactions.dm [hunk 9]
```diff
@@ -193,44 +193,43 @@
...
-	cached_gases[/datum/gas/nitrous_oxide][MOLES] += waste_added
-	cached_gases[/datum/gas/nitryl][MOLES] += waste_added
-	cached_gases[/datum/gas/carbon_dioxide][MOLES] += waste_added
-	if (location)
-		radiation_pulse(location, reaction_energy/(PLASMA_BINDING_ENERGY*MAX_CATALYST_EFFICENCY))
+	var/mediation = (air.heat_capacity()-(cached_gases[/datum/gas/plasma][MOLES]*cached_gases[/datum/gas/plasma][GAS_META][META_GAS_SPECIFIC_HEAT]))/(air.total_moles()-cached_gases[/datum/gas/plasma][MOLES])
+	var/carbon_amount = 2*cached_gases[/datum/gas/carbon_dioxide][MOLES]
+	var/plasma_fused = 0
+	if (carbon_amount < 20) //Lack of catalyst, fusion reaction starts to break down.
+		plasma_fused = carbon_amount
```
Comment (duncathan) [inline code/modules/atmospherics/gasmixtures/reactions.dm]: you're aware that no plasma is actually fused in this branch, correct?

**Review by duncathan — CHANGES_REQUESTED** [review]

File: code/modules/atmospherics/gasmixtures/reactions.dm (hunk 2 above)
Comment (as334) [reply code/modules/atmospherics/gasmixtures/reactions.dm]: Average heat capacity not including plasma, but yeah I will.

File: code/modules/atmospherics/gasmixtures/reactions.dm (hunk 7 above)
Comment (as334) [reply code/modules/atmospherics/gasmixtures/reactions.dm]: Pretty much, although it's occurred to me I need to set the reaction threshold for CO2 higher to make it work.

File: code/modules/atmospherics/gasmixtures/reactions.dm (hunk 6 above)
Comment (as334) [reply code/modules/atmospherics/gasmixtures/reactions.dm]: Yeah

File: code/modules/atmospherics/gasmixtures/reactions.dm (hunk 8 above)
Comment (as334) [reply code/modules/atmospherics/gasmixtures/reactions.dm]: I don't believe so? It's fine if the reaction uses up all the available gas, so going to 0 should be alright.

File: code/modules/atmospherics/gasmixtures/reactions.dm (hunk 5 above)
Comment (duncathan) [reply code/modules/atmospherics/gasmixtures/reactions.dm]: ah my bad it's mol*mol/(J/K) = mol which still doesn't make sense

**Comment by Buggy123** [comment]
So on a scale of 1 to 1000/2000/4000, how goes is this new pulse fusion for bombmaking?

File: code/modules/atmospherics/gasmixtures/reactions.dm (hunk 3 above)
Comment (praisenarsie) [reply code/modules/atmospherics/gasmixtures/reactions.dm]: 2 is better than 1!!!

**Comment by praisenarsie** [comment]
this is gonna hide behind the bombcap afaik unless there's a check somewhere it will ignore the bombcap.

**Comment by duncathan** [comment]
as334 and I may know more about the inner workings of atmos than most, but we're not great at powergaming things like bomb making. it'll be up to the real powergamers to play with this and see.

**Comment by CPTANT** [comment]
So at what temperature and pressure will fusion occur and what temperature does it reach?

File: code/modules/atmospherics/gasmixtures/reactions.dm (hunk 2 above)
Comment (Supermichael777) [reply code/modules/atmospherics/gasmixtures/reactions.dm]: Clamp this to prevent negative numbers and zeros, you don't want division to fuck up

**Comment by Supermichael777** [comment]
I dont think this is posible in TTV bombs. have a test setup for it though

**Comment by Supermichael777** [comment]
also your maths are.... unit agnostic

File: code/modules/atmospherics/gasmixtures/reactions.dm (hunk 2 above)
Comment (as334) [reply code/modules/atmospherics/gasmixtures/reactions.dm]: Those shouldn't be possible, since that would either requiring total_moles to be less than the amount of plasma moles(impossible, equal at most), or require them to be the same, which would cause the reaction to not meet its minimum requirement for carbon dioxide for the reaction to occur.

Comment (Supermichael777) [reply code/modules/atmospherics/gasmixtures/reactions.dm]: Its just that this has been comically broken before, the math on this one is hard to follow across the 4 pages

Comment (as334) [reply code/modules/atmospherics/gasmixtures/reactions.dm]: While fusion does have an impressive tendency to break in crazy ways, I think that even if it's possible for a bug like that to occur here, we're better off seeing it break then hiding it with a clamp or a max. Since that would imply either reaction checking is broken(pretty bad) or total_moles() is broken(REALLY bad). As it stands though I think it's fine as it is

File: code/modules/atmospherics/gasmixtures/reactions.dm (hunk 1 above)
Comment (as334) [reply code/modules/atmospherics/gasmixtures/reactions.dm]: I'm predicting in the far flung future this comment will actually make up the majority of reactions.dm due to continuous fusion reworks.

File: code/modules/atmospherics/gasmixtures/reactions.dm (hunk 2 above)
Comment (duncathan) [reply code/modules/atmospherics/gasmixtures/reactions.dm]: this formula runtiming would be my worst nightmare lol, this is quite safe

**Comment by kevinz000** [comment]
please make sure this doesnt' break the server 🤗

**Comment by Supermichael777** [comment]
This is not currently possible to weaponize in tanks

**Comment by Supermichael777** [comment]
Not posible without a burntank heater

**Comment by CPTANT** [comment]
Does it still give off radiation?

**Comment by arsserpentarium** [comment]
I suggest to make fusion depends on emitter beam or explosion.So it goes only when there is emitter beam on tile. This way we can build fairly realistic fusion reactor, when heat and radiation is produced by igniting plasma with laser beam.
Also you can make with regular plasma bomb and fusion cannister H-bomb with great destructive power.

**Comment by KorPhaeron** [comment]
Stale, ping for reopen, etc
